[decision]: Changetracking (round 2) (#4714)
ElektraInitiative/libelektra: doc/decisions/1_in_discussion/change_tracking.md

## Solutions - Query

### Provide an API within `libelektra-kdb`

The API should be useable both by plugins and applications utilizing ELektra.
The API may look something like this:

```c
bool elektraChangeTrackingIsEnabled (KDB * kdb);
ChangeTrackingContext * elektraChangeTrackingGetContext (KDB * kdb, Key * parentKey);

KeySet * elektraChangeTrackingGetAddedKeys (ChangeTrackingContext * context);
KeySet * elektraChangeTrackingGetRemovedKeys (ChangeTrackingContext * context);
KeySet * elektraChangeTrackingGetModifiedKeys (ChangeTrackingContext * context);

bool elektraChangeTrackingValueChanged (ChangeTrackingContext * context, Key * key);
bool elektraChangeTrackingMetaChanged (ChangeTrackingContext * context, Key * key);

KeySet * elektraChangeTrackingGetAddedMetaKeys (ChangeTrackingContext * context, Key * key);
KeySet * elektraChangeTrackingGetRemovedMetaKeys (ChangeTrackingContext * context, Key * key);
KeySet * elektraChangeTrackingGetModifiedMetaKeys (ChangeTrackingContext * context, Key * key);

Key * elektraChangeTrackingGetOriginalKey (ChangeTrackingContext * context, Key * key);

> const Key * elektraChangeTrackingGetOriginalMetaKey (ChangeTrackingContext * context, Key * key, const char * metaName);

Reviewer: Please add some context why this would be needed (connect to use case etc.). Isn't three KeySets with all added/removed/modified (meta)keys enough?

Author: Good catch! Yes, if `elektraChangeTrackingGetModifiedKeys` / `elektraChangeTrackingGetModifiedMetaKeys` return the *old* keys, then these functions are not required.

- if overhead is not negligible: only do tracking as required, i.e., a plugin specifically requests it
- have negligible overhead if disabled
- not duplicate data for each plugin that wants change tracking
- work with all allowed sequences of `kdbGet` and `kdbSet` as [per this decision](../0_drafts/operation_sequences.md)

We only want to track changes that are done by the user, not changes that are done by plugins.
I.e. the scope of change tracking is on what happens _outside_ of `kdbGet` and `kdbSet`.

The library `libelektra-core` must be kept minimal.

## Assumptions

- It is possible to do change tracking with reasonable memory and computation overhead
- It is possible to design a single change tracking API that is useful for all existing and future plugins
- False positivies are okay

>   - this may happend when some keys may have been changed by the user, but have subsequentially been "unchanged" by a transformation plugin

Reviewer 2: ```suggestion
  - this may happend when some keys have been changed by the user, but have subsequentially been "unchanged" by a transformation plugin
```

    Scenario: plugin that converts `false`<->`0` and `true`<->`1`
    - `system:/background` is stored with value `false`
    - user gets key `system:/background` with value `0` (after conversion by plugin)
    - user changes it to `false`
    - changetracking detects that value has been changed, because `false` != `0`
    - plugin changes value from `false` to `0`

>     - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed

Reviewer 2: ```suggestion
    - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed
    - We assume consumers of the changetracking API have access to both the previous and the new value of a changed key.
      Therefore, they could detect these false positive cases, if the need to.
```

- not duplicate data for each plugin that wants change tracking
- work with all allowed sequences of `kdbGet` and `kdbSet` as [per this decision](operation_sequences.md)

We only want to track changes that are done by the user, not changes that are done by plugins.
I.e. the scope of change tracking is on what happens _outside_ of `kdbGet` and `kdbSet`.

The library `libelektra-core` must be kept minimal.

## Assumptions

- It is possible to do change tracking with reasonable memory and computation overhead
- It is possible to design a single change tracking API that is useful for all existing and future plugins
- False positivies are okay

  - this may happend when some keys may have been changed by the user, but have subsequentially been "unchanged" by a transformation plugin

>     Scenario: hypothetical plugin that converts all values to UPPERCASE.
>     - user gets key `system:/background` with value `RED`
>     - user changes it to `red`
>     - changetracking detects that value has been changed, because `RED` != `red`
>     - plugin changes value to UPPERCASE `RED`
>     - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed

Reviewer 2: Let's use something that really exists

```suggestion
    Scenario: `type` plugin configured for boolean conversion with `false`/`true`
    - `system:/background` is stored with value `false`
    - user gets key `system:/background` with value `0` (after conversion by `type`)
    - user changes it to `false`
    - changetracking detects that value has been changed, because `false` != `0`
    - `type` plugin changes value to `0`
    - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed
```

Author: Funny, in #4621 it was suggested that we don't name a plugin specifically 😅

Reviewer 2: Hm, I guess there are pros and cons. I just think giving it a name makes this example more readable. I wouldn't even have to be a specific existing plugin. Also I think the `false` -> `0` conversion would be much more common and therefore it would be easier to understand why this is useful/important.

So maybe we go with:

```suggestion
    Scenario: plugin that converts `false`<->`0` and `true`<->`1`
    - `system:/background` is stored with value `false`
    - user gets key `system:/background` with value `0` (after conversion by plugin)
    - user changes it to `false`
    - changetracking detects that value has been changed, because `false` != `0`
    - plugin changes value from `false` to `0`
    - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed
```

Note also the additional line at the beginning explaining how the value is stored.

- False negatives are not okay

## Considered Alternatives

> ### Alternative 1 - Tracking withn `libelektra-kdb`

Reviewer: ```suggestion
### Alternative 1 - Tracking within `libelektra-kdb`
```

Do the tracking within `libelektra-kdb`, within the `kdbGet` and `kdbSet` operations.

> Essentially, we'd have a 'giant' keyset for all keys that were used returned by a `kdbGet` operation in the lifetime of the `kdb` instance.

Reviewer: ```suggestion
Essentially, we'd have an internal cache for all keys that were used returned by a `kdbGet` operation in the lifetime of the `kdb` instance.
```

Use the `backendData->keys` for change tracking

We already store which keys have been returned by `kdbGet` for each backend within KDB.

> Currently, however, this is not a deep copy, as we are returning the keys from `backendData->keys` directly.

Reviewer: `backendData->keys` is an implementation detail. Is it the internal cache or something else?

Author: If we leave out this detail, then yes it would probably be the same as Alternative 1. This distinction was made before copy-on-write, as `backendData->keys` already stores the keys returned, whereas for Alternative 1, we'd have to have another duplication.

With copy-on-write, this distinction gets much more fuzzy.

Reviewer 2: ```suggestion
Currently, however, this is not a deep copy, as we are returning the internally stored `Key` instances directly.
```

We don't need to say where they are stored, just that the keys are returned and therefore can be changed externally.

This means we can not detect changes to the values or metadata of keys right now.
We can, however, rely on this for detecting removed and added keys in its current form.

As we now implement copy-on-write for all keys and keysets within Elektra, we can deep copy what we return in `kdbGet`.

> Another problem with this approach is that `backendData->keys` get cleared every time `kdbGet` is called.

Reviewer: Again, please avoid internals like `backendData->keys` (or introduce them, as not everyone knows the details of new-backend yet).

Reviewer 2: ```suggestion
Another problem with this approach is that the internally stored keys are recreated as new instances every time `kdbGet` is called.
```

Not quite true I think, but good enough here.
